src/AstTranslator.h

```diff
@@ -40,44 +48,266 @@ class TypeEnvironment;
+            setRecordDefinition(init, Location({level, pos, std::move(name)}));
+        }
+
+        const Location& getDefinitionPoint(const AstRecordInit& init) const {
+            auto pos = record_definitions.find(&init);
+            if (pos != record_definitions.end()) {
+                return pos->second;
+            }
+            assert(false && "Requested location for undefined record!");
+
+            static Location fail;
+            return fail;
+        }
+
+        // - unpacking -
+
+        void setRecordUnpackLevel(const AstRecordInit& init, int level) {
+            record_unpacks[&init] = level;
+        }
+
+        int getRecordUnpackLevel(const AstRecordInit& init) const {
+            auto pos = record_unpacks.find(&init);
+            if (pos != record_unpacks.end()) {
+                return pos->second;
+            }
+            assert(false && "Requested record is not unpacked properly!");
+            return 0;
+        }
+
+        // -- aggregates --
+
+        void setAggregatorLocation(const AstAggregator& agg, const Location& loc) {
+            aggregator_locations.push_back(std::make_pair(&agg, loc));
+        }
+
+        const Location& getAggregatorLocation(const AstAggregator& agg) const {
+            // search list
+            for (const auto& cur : aggregator_locations) {
+                if (*cur.first == agg) {
+                    return cur.second;
+                }
+            }
+
+            // fail
+            std::cout << "Lookup of " << &agg << " = " << agg << " failed\n";
+            assert(false && "Requested aggregation operation is not processed!");
+
+            const static Location fail = Location();
+            return fail;
+        }
+
+        // -- others --
+
+        bool isSomethingDefinedOn(int level) const {
+            // check for variable definitions
+            for (const auto& cur : var_references) {
+                if (cur.second.begin()->level == level) {
+                    return true;
+                }
+            }
+            // check for record definitions
+            for (const auto& cur : record_definitions) {
+                if (cur.second.level == level) {
+                    return true;
+                }
+            }
+            // nothing defined on this level
+            return false;
+        }
+
+        void print(std::ostream& out) const {
+            out << "Variables:\n\t";
+            out << join(var_references, "\n\t");
+        }
+
+        friend std::ostream& operator<<(std::ostream& out, const ValueIndex& index) __attribute__((unused)) {
+            index.print(out);
+            return out;
+        }
+    };
+
+    /**
+     * A utility function assigning names to unnamed variables such that enclosing
+     * constructs may be cloned without losing the variable-identity.
+     */
+    void nameUnnamedVariables(AstClause* clause);
```

Comment:
Do all these need to be global?

Reply:
Yes - we need context and we don't want to return to a style where we pass on global state of the AstTranslator object  via parameters.
The functions were converted to methods. 
We still need to tidy up AstTranslator in a later stage.